### PR TITLE
Add a DiagnosticBuilder to support context on diagnostics.

### PR DESCRIPTION
**Pull request by jonmeow**
This is currently used once for PreviousDefinition in semantics.

This PR doesn't just add a builder, it also adds support to the emitter itself to collect notes attoched to a diagnostic, and to the consumers and emitters to print all of them.

### CODE REVIEW

File: toolchain/diagnostics/diagnostic_emitter.h [hunk 1]
```diff
@@ -155,6 +181,55 @@ struct DiagnosticBase {
 template <typename LocationT>
 class DiagnosticEmitter {
  public:
+  class DiagnosticBuilder {
```
Comment (chandlerc) [inline toolchain/diagnostics/diagnostic_emitter.h:187]: Could you add some comments for the class and its public API?

Also, would `Build` be a reasonable name given that its a nested class?

File: toolchain/diagnostics/diagnostic_emitter.h [hunk 2]
```diff
@@ -155,6 +181,55 @@ struct DiagnosticBase {
...
+   public:
+    template <typename... Args>
+    auto WithContext(
```
Comment (chandlerc) [inline toolchain/diagnostics/diagnostic_emitter.h]: Maybe `Note` (the verb) or `NoteContext` (depending on the higher up question) to help remind readers that it will be emitted as a note diagnostic?

File: toolchain/diagnostics/diagnostic_emitter.h [hunk 3]
```diff
@@ -79,6 +102,9 @@ struct Diagnostic {
 
   // Returns the formatted string. By default, this uses llvm::formatv.
   std::function<std::string(const Diagnostic&)> format_fn;
+
+  // Context for the diagnostic.
+  llvm::SmallVector<std::unique_ptr<Diagnostic>> context;
```
Comment (chandlerc) [inline toolchain/diagnostics/diagnostic_emitter.h]: Here and throughout -- what do you think about calling these "notes" as opposed to "context"?

For me, "notes" focuses better on what they are, while "context" is a (possibly) reason. There might be other reasons we want to attach notes beyond context.

File: toolchain/diagnostics/diagnostic_emitter.h [hunk 4]
```diff
@@ -168,20 +243,23 @@ class DiagnosticEmitter {
...
-        .format_fn = [&diagnostic_base](const Diagnostic& diagnostic)
-            -> std::string { return diagnostic_base.FormatFn(diagnostic); },
-    });
+            typename std::common_type_t<Args>... args) -> void {
+    DiagnosticBuilder(this, location, diagnostic_base, std::move(args)...)
+        .Emit();
+  }
+
+  template <typename... Args>
+  auto Build(LocationT location,
```
Comment (chandlerc) [inline toolchain/diagnostics/diagnostic_emitter.h:269]: Seems good to have some documentation here, especially maybe an example since this is a fluent-API where that might be more important than other kinds of APIs?

File: toolchain/diagnostics/diagnostic_emitter.h [hunk 5]
```diff
@@ -87,7 +113,7 @@ class DiagnosticConsumer {
   virtual ~DiagnosticConsumer() = default;
 
   // Handle a diagnostic.
-  virtual auto HandleDiagnostic(const Diagnostic& diagnostic) -> void = 0;
+  virtual auto HandleDiagnostic(Diagnostic diagnostic) -> void = 0;
```
Comment (chandlerc) [inline toolchain/diagnostics/diagnostic_emitter.h:116]: Why this change? Seems somewhat wasteful to copy (or even move) the diagnostic objects here... They're going to end up quite large with `std::function` objects and small vectors, etc.

I could also imagine wanting to support multiple consumers directly (instead of the chaining done by wrapping consumers below), but that's maybe not relevant yet.

**Review by chandlerc — CHANGES_REQUESTED** [review]
Very cool! Comments inline.

One thing that I'd also like to ask for is a bit more context (ba-dum) / summary of the overall change in the PR/commit description if you don't mind. It just helps me get everything in my head before i start reading the code.

For example, this PR doesn't just add a builder, it also adds support to the emitter itself to collect notes attoched to a diagnostic, and to the consumers and emitters to print all of them.

**Comment by jonmeow** [comment]
> Very cool! Comments inline.
> 
> One thing that I'd also like to ask for is a bit more context (ba-dum) / summary of the overall change in the PR/commit description if you don't mind. It just helps me get everything in my head before i start reading the code.
> 
> For example, this PR doesn't just add a builder, it also adds support to the emitter itself to collect notes attoched to a diagnostic, and to the consumers and emitters to print all of them.

Copied into description.

File: toolchain/diagnostics/diagnostic_emitter.h (hunk 3 above)
Comment (jonmeow) [reply toolchain/diagnostics/diagnostic_emitter.h]: Switched to "notes"

File: toolchain/diagnostics/diagnostic_emitter.h (hunk 1 above)
Comment (jonmeow) [reply toolchain/diagnostics/diagnostic_emitter.h:187]: Wouldn't `Build` be a name conflict with the `Build` function?

I meant, I prefer a longer name for explicitness. But I think it's also helpful for unique names. And I don't think it matters _too_ much since users wouldn't right this.

File: toolchain/diagnostics/diagnostic_emitter.h (hunk 4 above)
Comment (jonmeow) [reply toolchain/diagnostics/diagnostic_emitter.h:269]: Done

File: toolchain/diagnostics/diagnostic_emitter.h (hunk 2 above)
Comment (jonmeow) [reply toolchain/diagnostics/diagnostic_emitter.h]: Done (Note), not sure what you mean by "higher up question".

File: toolchain/diagnostics/diagnostic_emitter.h (hunk 5 above)
Comment (jonmeow) [reply toolchain/diagnostics/diagnostic_emitter.h:116]: The consequence of using a const& here is that the SortingDiagnosticConsumer needs to copy, since it can't move. Do you prefer the copy instead of moves?

File: toolchain/diagnostics/diagnostic_emitter.h [hunk 6]
```diff
@@ -79,6 +102,9 @@ struct Diagnostic {
 
   // Returns the formatted string. By default, this uses llvm::formatv.
   std::function<std::string(const Diagnostic&)> format_fn;
+
+  // Context for the diagnostic.
```
Comment (chandlerc) [inline toolchain/diagnostics/diagnostic_emitter.h]: Maybe:

```suggestion
  // Notes that add context or supplemental information to the diagnostic.
```

File: toolchain/diagnostics/diagnostic_emitter.h (hunk 5 above)
Comment (chandlerc) [reply toolchain/diagnostics/diagnostic_emitter.h:116]: Hmm, that's an interesting question.

But that seems like an unrelated change to the rest of this PR?

FWIW, I'd be curious about a model where we ensured the diagnostics lived until after `Flush` is called so that the sorted consumer can actually use pointers to the stable diagnostic locations. But not sure any of this matters enough to do too much churn at this stage. Seems like we can revisit any of these decisions w/ performance data if we needed. I mostly was surprised to see the *change* here.

What seems best as a way forward? My inclination is to leave this API and revisit it later, but happy to do something else if that seems better or if that would just require a bunch of other changes to make the new notes stuff work well.

File: toolchain/diagnostics/diagnostic_emitter.h (hunk 1 above)
Comment (chandlerc) [reply toolchain/diagnostics/diagnostic_emitter.h:187]: Sorry, I meant `Builder` instead of `DiagnsoticBuilder`.

Anyways, not worried about the name much, whatever works for you. I'm suggesting some additional information and detail in the comments that would help me out as a reader of the API and code.

File: toolchain/diagnostics/diagnostic_emitter.h [hunk 7]
```diff
@@ -155,6 +181,57 @@ struct DiagnosticBase {
 template <typename LocationT>
 class DiagnosticEmitter {
  public:
+  // Builds a diagnostic for emitting. See `DiagnosticEmitter::Build` for use.
```
Comment (chandlerc) [inline toolchain/diagnostics/diagnostic_emitter.h]:
```suggestion
  // A builder-pattern type to provide a fluent interface for constructing
  // a more complex diagnostic. See `DiagnosticEmitter::Build` for the
  // expected usage.
```

File: toolchain/diagnostics/diagnostic_emitter.h [hunk 8]
```diff
@@ -155,6 +181,57 @@ struct DiagnosticBase {
...
+  class DiagnosticBuilder {
+   public:
+    // Adds a note. See `DiagnosticEmitter::Build` for use.
```
Comment (chandlerc) [inline toolchain/diagnostics/diagnostic_emitter.h]:
```suggestion
    // Adds a note diagnostic attached to the main diagnostic being built.
    // The API mirrors the main emission API: `DiagnosticEmitter::Emit`.
    // For the expected usage see the builder API: `DiagnosticEmitter::Build`.
```

File: toolchain/diagnostics/diagnostic_emitter.h [hunk 9]
```diff
@@ -155,6 +181,57 @@ struct DiagnosticBase {
...
+          emitter_->translator_->GetLocation(location), diagnostic_base.Format,
+          llvm::SmallVector<llvm::Any, 0>({std::move(args)...}),
+          [&diagnostic_base](const Diagnostic& diagnostic) -> std::string {
+            return diagnostic_base.FormatFn(diagnostic);
+          },
+          llvm::SmallVector<std::unique_ptr<Diagnostic>>()));
+      return *this;
+    }
+
+    // Emits the diagnostic. See `DiagnosticEmitter::Build` for use.
```
Comment (chandlerc) [inline toolchain/diagnostics/diagnostic_emitter.h]:
```suggestion
    // Emits the built diagnostic and its attached notes.
    // For the expected usage see the builder API: `DiagnosticEmitter::Build`.
```

File: toolchain/diagnostics/diagnostic_emitter.h [hunk 10]
```diff
@@ -168,20 +245,28 @@ class DiagnosticEmitter {
...
+            typename std::common_type_t<Args>... args) -> void {
+    DiagnosticBuilder(this, location, diagnostic_base, std::move(args)...)
+        .Emit();
+  }
+
+  // Builds a diagnostic to emit. For example:
+  //
+  //   emitter_.Build(location1, MyDiagnostic)
+  //     .Note(location2, MyDiagnosticNote)
+  //     .Emit()
```
Comment (chandlerc) [inline toolchain/diagnostics/diagnostic_emitter.h]:
```suggestion
  // A fluent interface for building a diagnostic and attaching notes for added
  // context or information. For example:
  //
  //   emitter_.Build(location1, MyDiagnostic)
  //     .Note(location2, MyDiagnosticNote)
  //     .Emit();
```

**Review by chandlerc — APPROVED** [review]
Mostly some extra details for comments and a question about how to handle the consumer API in this PR. Somewhat happy with either approach for now, so feel free to merge w/ whichever works as long as my comment adjustments make sense to you. Let me know if another look would help though.

File: toolchain/diagnostics/diagnostic_emitter.h (hunk 5 above)
Comment (jonmeow) [reply toolchain/diagnostics/diagnostic_emitter.h:116]: It's an issue on this PR because the implicit copy no longer works -- pointers can't be copied. So I can either implement a custom copy that duplicates the pointers, or do moves. Either way I need to implement something here. I don't think "leave this API" is an option.

1. Use shared_ptr instead of unique_ptr (ambiguous ownership issue)
2. custom copy that duplicates unique_ptr Diagnostics
3. move

Comment (chandlerc) [reply toolchain/diagnostics/diagnostic_emitter.h:116]: Ah, the fact that we now have `unique_ptr` (and not a more `Box` like construct) is what eluded me, thanks!

I'm happy moving for now, even if it makes sense to revisit it later.

Comment (jonmeow) [reply toolchain/diagnostics/diagnostic_emitter.h:116]: Per discussion, will send a followup that:

- Documents why we move more explicitly.
- Switch to a wrapper class to try getting rid of unique_ptr (but still not copying).
